### DISCUSSION TITLE
Use custom component inside remark plugin & return component

### DISCUSSION THREAD

**Discussion by ZenaMel** [facebook/docusaurus#8988] [Q&A] [answered]
What I'm trying to accopmlish:

I've got a custom component `<DeepLink />`. Instead of making writers write `<DeepLink href="<URI>" />`, I'd like to provide a shorthand, e.g.,

```
deeplink::192.168.123.123
```

Here's the current remark plugin code:

```js
const visit = require('unist-util-visit');
import { DeepLink } from '../components/DeepLink';

/*
Usage: deeplink::192.168.123.123 inside md files.
*/

const plugin = (options) => {
  const transformer = async (ast) => {
    visit(ast, 'text', (node) => {
      node.value = node.value.replace(/deeplink::*\S+/g, (match) => {
        link = match.replace('deeplink::', '');
        link = "http://" + link;
        return <DeepLink href="${link}" />;
      });
    });
  };
  return transformer;
};

module.exports = plugin;
```

The code for that doesn't work yet as there are problems I'm unsure on how to fix:

- import { DeepLink } from '../components/DeepLink'; leads to --> SyntaxError: Cannot use import statement outside a module
- returning `<DeepLink />` like I'm trying to doesn't work --> SyntaxError: Unexpected token '<'

**Comment by slorber** [comment]
With mdx1 you should return the string of your component instead of returning a  React element. You should also make sure your component is imported. 

In mdx2 this is quite different and you should emit some esast AST nodes. There's no quick win, you really have to learn how all this AST processing work by reading Unified and MDX docs.

You can also take inspiration from our existing plugins that do similar things like the admonition plugin. My suggestion is to implement this for mdx2  directly and use the remark-directive syntax + do like we did for admonitions, that will be simpler than inventng your own syntax.

Doing something quick and dirty is less likely to work in mdx2, so you have 2 choices: you learn how it works, or you don't do it 😅

**Reply by ZenaMel** [reply]
Do I have the _option_ in docusaurus 2.4.1 to either work with mdx v1 or v2? Or has everything been migrated to mdx v2?

**Reply by slorber** [reply, marked as answer]
Docusaurus v2 is MDX v1
Docusaurus v3 will be MDX v2

The MDX upgrade is a significant change because the AST changed, and most plugins (including ours) had to be completely rewritten. It will, unfortunately, be the case for our community too. The good news is that now MDX is much better, stable and less likely to change again in such a disruptive way in the future.

For this reason, you should rather implement this for MDX v2 / Docusaurus v3 directly
